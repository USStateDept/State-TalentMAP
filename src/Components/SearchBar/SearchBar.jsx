import { Component } from 'react';
import PropTypes from 'prop-types';
import FA from 'react-fontawesome';
import { EMPTY_FUNCTION, PREVENT_DEFAULT } from 'Constants/PropTypes';
import InteractiveElement from '../InteractiveElement';
import { focusById } from '../../utilities';

class SearchBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchText: { value: this.props.defaultValue || '' },
    };
  }

  changeText = e => {
    const { searchText } = this.state;
    searchText.value = e.target.value;
    this.setState({ searchText }, this.props.onChangeText(e));
  };

  clearSearch = () => {
    const { id } = this.props;
    this.changeText({ target: { value: '' } });
    this.props.onClear();
    focusById(id, 1);
  };

  render() {
    const hidden = {
      display: 'none',
    };
    const { id, type, submitDisabled, submitText, placeholder, inputDisabled,
      alertText, onSubmitSearch, label, labelSrOnly, noForm, noButton, showClear }
      = this.props;
    const { searchText } = this.state;
    let showSubmitText = true; // do not hide submit text initially
    if (type === 'small') { showSubmitText = false; } // small search class should not have text

    let labelClass = '';
    if (labelSrOnly) { labelClass = 'usa-sr-only'; }

    let formattedSubmitText = '';
    let formattedSubmitTextSr = 'Search';
    if (showSubmitText) {
      formattedSubmitText = submitText;
      formattedSubmitTextSr = submitText;
    }

    let submitDisabledClass = '';
    if (submitDisabled) {
      submitDisabledClass = 'usa-button-disabled';
    }

    const hasValue = !!searchText.value;

    const input = (
      <input
        id={id}
        value={inputDisabled ? '' : searchText.value}
        onChange={this.changeText}
        type="search"
        name="search"
        placeholder={placeholder}
        disabled={inputDisabled}
      />
    );
    const child = (
      <div className="usa-grid-full label-input-wrapper">
        <label className={labelClass} htmlFor={id}>
          {label}
        </label>
        {
          showClear ?
            <span className="text-input-wrapper">
              {input}
              {
                hasValue ? (
                  <InteractiveElement onClick={this.clearSearch} type="span" role="button" title="Clear keyword">
                    <FA name="times-circle" />
                  </InteractiveElement>
                ) : <FA name="search" />
              }
            </span>
            :
            input
        }
        <div id={`enabled-search-${id}`}>
          { !noButton &&
          <button
            id={`enabled-search-button-${id}`}
            className={submitDisabledClass}
            disabled={submitDisabled}
            type="submit"
            title="submit search"
          >
            <span className="usa-search-submit-text">{formattedSubmitText}</span>
            <span className="usa-sr-only">{formattedSubmitTextSr}</span>
          </button>
          }
        </div>
        <div id={`disabled-search-${id}`} style={hidden}>
          {
            !noButton &&
            <button
              className="usa-button-disabled"
              disabled="true"
              type="submit"
              id={`disabled-search-button-${id}`}
              title="search button disabled"
            >
              <span className="usa-search-submit-text usa-button-disabled">
                {formattedSubmitText}
              </span>
            </button>
          }
          <span className="alert-text">{alertText}</span>
        </div>
      </div>
    );
    return (
      <div className={`usa-search usa-search-${type} searchbar`}>
        <div role="search" className="usa-grid-full">
          { !noForm &&
            <form onSubmit={onSubmitSearch}>
              {child}
            </form>
          }
          {
            noForm &&
            child
          }
        </div>
      </div>
    );
  }
}

SearchBar.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.node,
  type: PropTypes.oneOf(['small', 'medium', 'big']),
  submitDisabled: PropTypes.bool,
  submitText: PropTypes.string.isRequired,
  alertText: PropTypes.string,
  onChangeText: PropTypes.func,
  onSubmitSearch: PropTypes.func,
  labelSrOnly: PropTypes.bool,
  noForm: PropTypes.bool,
  noButton: PropTypes.bool,
  placeholder: PropTypes.string,
  defaultValue: PropTypes.string,
  inputDisabled: PropTypes.bool,
  showClear: PropTypes.bool,
  onClear: PropTypes.func,
};

SearchBar.defaultProps = {
  type: 'big', // should be one of the USWDS search types - https://standards.usa.gov/components/search-bar/
  submitDisabled: false,
  alertText: 'Disabled',
  label: 'Search', // sr only if flagged
  labelSrOnly: true,
  noForm: false,
  noButton: false,
  placeholder: null,
  defaultValue: null,
  onSubmitSearch: PREVENT_DEFAULT,
  onChangeText: EMPTY_FUNCTION,
  inputDisabled: false,
  showClear: false,
  onClear: EMPTY_FUNCTION,
};

export default SearchBar;
