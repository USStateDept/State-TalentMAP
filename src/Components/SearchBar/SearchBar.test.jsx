import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import SearchBar from './SearchBar';

['small', 'medium', 'big'].forEach((size) => {
  describe(`SearchBarComponent ${size}`, () => {
    let wrapper = null;

    it('can receive props', () => {
      wrapper = shallow(
        <SearchBar
          id="search"
          type={size}
          submitDisabled
          submitText="Submit"
          alertText="Search disabled"
          onChangeText={() => {}}
          onSubmitSearch={() => {}}
        />,
      );
      expect(wrapper.instance().props.id).toBe('search');
    });

    it('handles different props', () => {
      wrapper = shallow(
        <SearchBar
          id="search-2"
          label="Label"
          type={size}
          submitDisabled={false}
          submitText="Submit 2"
          alertText="Search is disabled"
          onChangeText={() => {}}
          onSubmitSearch={() => {}}
        />,
      );
      expect(wrapper.instance().props.id).toBe('search-2');
    });

    it('can change text', () => {
      wrapper = shallow(
        <SearchBar
          id="search-2"
          type={size}
          submitDisabled={false}
          submitText="Submit 2"
          alertText="Search is disabled"
          onChangeText={() => {}}
          onSubmitSearch={() => {}}
        />,
      );
      wrapper.find('input').simulate('change', { target: { value: 'test' } });
      expect(wrapper.instance().state.searchText.value).toBe('test');
    });

    it('can submit the form', () => {
      const spy = sinon.spy();
      wrapper = shallow(
        <SearchBar
          id="search-2"
          type={size}
          submitDisabled={false}
          submitText="Submit 2"
          alertText="Search is disabled"
          onChangeText={() => {}}
          onSubmitSearch={spy}
        />,
      );
      wrapper.find('input').simulate('change', { target: { value: 'test' } });
      wrapper.find('form').simulate('submit');
      expect(spy.calledOnce).toBe(true);
    });

    it('sets state and calls props.onClear() on this.onClear()', () => {
      const spy = sinon.spy();
      wrapper = shallow(<SearchBar
        id="search-2"
        type={size}
        submitDisabled={false}
        submitText="Submit 2"
        alertText="Search is disabled"
        onChangeText={() => {}}
        onSubmitSearch={() => {}}
        showClear
        onClear={spy}
      />);
      wrapper.instance().setState({ searchText: { value: 'abc' } });
      wrapper.find('InteractiveElement').simulate('click');
      expect(wrapper.instance().state.searchText.value).toBe('');
      sinon.assert.calledOnce(spy);
    });

    it('matches snapshot', () => {
      wrapper = shallow(
        <SearchBar
          id="search-2"
          label="Label"
          type={size}
          submitDisabled={false}
          submitText="Submit 2"
          alertText="Search is disabled"
          onChangeText={() => {}}
          onSubmitSearch={() => {}}
        />,
      );
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('matches snapshot when showClear === true', () => {
      wrapper = shallow(
        <SearchBar
          id="search-2"
          label="Label"
          type={size}
          submitDisabled={false}
          submitText="Submit 2"
          alertText="Search is disabled"
          onChangeText={() => {}}
          onSubmitSearch={() => {}}
          showClear
        />,
      );
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });
});
