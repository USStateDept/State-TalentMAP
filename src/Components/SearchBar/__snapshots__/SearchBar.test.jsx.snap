// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBarComponent big matches snapshot 1`] = `
<div
  className="usa-search usa-search-big searchbar"
>
  <div
    className="usa-grid-full"
    role="search"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="usa-grid-full label-input-wrapper"
      >
        <label
          className="usa-sr-only"
          htmlFor="search-2"
        >
          Label
        </label>
        <input
          disabled={false}
          id="search-2"
          name="search"
          onChange={[Function]}
          placeholder={null}
          type="search"
          value=""
        />
        <div
          id="enabled-search-search-2"
        >
          <button
            className=""
            disabled={false}
            id="enabled-search-button-search-2"
            title="submit search"
            type="submit"
          >
            <span
              className="usa-search-submit-text"
            >
              Submit 2
            </span>
            <span
              className="usa-sr-only"
            >
              Submit 2
            </span>
          </button>
        </div>
        <div
          id="disabled-search-search-2"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <button
            className="usa-button-disabled"
            disabled="true"
            id="disabled-search-button-search-2"
            title="search button disabled"
            type="submit"
          >
            <span
              className="usa-search-submit-text usa-button-disabled"
            >
              Submit 2
            </span>
          </button>
          <span
            className="alert-text"
          >
            Search is disabled
          </span>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`SearchBarComponent big matches snapshot when showClear === true 1`] = `
<div
  className="usa-search usa-search-big searchbar"
>
  <div
    className="usa-grid-full"
    role="search"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="usa-grid-full label-input-wrapper"
      >
        <label
          className="usa-sr-only"
          htmlFor="search-2"
        >
          Label
        </label>
        <span
          className="text-input-wrapper"
        >
          <input
            disabled={false}
            id="search-2"
            name="search"
            onChange={[Function]}
            placeholder={null}
            type="search"
            value=""
          />
          <FontAwesome
            name="search"
          />
        </span>
        <div
          id="enabled-search-search-2"
        >
          <button
            className=""
            disabled={false}
            id="enabled-search-button-search-2"
            title="submit search"
            type="submit"
          >
            <span
              className="usa-search-submit-text"
            >
              Submit 2
            </span>
            <span
              className="usa-sr-only"
            >
              Submit 2
            </span>
          </button>
        </div>
        <div
          id="disabled-search-search-2"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <button
            className="usa-button-disabled"
            disabled="true"
            id="disabled-search-button-search-2"
            title="search button disabled"
            type="submit"
          >
            <span
              className="usa-search-submit-text usa-button-disabled"
            >
              Submit 2
            </span>
          </button>
          <span
            className="alert-text"
          >
            Search is disabled
          </span>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`SearchBarComponent medium matches snapshot 1`] = `
<div
  className="usa-search usa-search-medium searchbar"
>
  <div
    className="usa-grid-full"
    role="search"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="usa-grid-full label-input-wrapper"
      >
        <label
          className="usa-sr-only"
          htmlFor="search-2"
        >
          Label
        </label>
        <input
          disabled={false}
          id="search-2"
          name="search"
          onChange={[Function]}
          placeholder={null}
          type="search"
          value=""
        />
        <div
          id="enabled-search-search-2"
        >
          <button
            className=""
            disabled={false}
            id="enabled-search-button-search-2"
            title="submit search"
            type="submit"
          >
            <span
              className="usa-search-submit-text"
            >
              Submit 2
            </span>
            <span
              className="usa-sr-only"
            >
              Submit 2
            </span>
          </button>
        </div>
        <div
          id="disabled-search-search-2"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <button
            className="usa-button-disabled"
            disabled="true"
            id="disabled-search-button-search-2"
            title="search button disabled"
            type="submit"
          >
            <span
              className="usa-search-submit-text usa-button-disabled"
            >
              Submit 2
            </span>
          </button>
          <span
            className="alert-text"
          >
            Search is disabled
          </span>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`SearchBarComponent medium matches snapshot when showClear === true 1`] = `
<div
  className="usa-search usa-search-medium searchbar"
>
  <div
    className="usa-grid-full"
    role="search"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="usa-grid-full label-input-wrapper"
      >
        <label
          className="usa-sr-only"
          htmlFor="search-2"
        >
          Label
        </label>
        <span
          className="text-input-wrapper"
        >
          <input
            disabled={false}
            id="search-2"
            name="search"
            onChange={[Function]}
            placeholder={null}
            type="search"
            value=""
          />
          <FontAwesome
            name="search"
          />
        </span>
        <div
          id="enabled-search-search-2"
        >
          <button
            className=""
            disabled={false}
            id="enabled-search-button-search-2"
            title="submit search"
            type="submit"
          >
            <span
              className="usa-search-submit-text"
            >
              Submit 2
            </span>
            <span
              className="usa-sr-only"
            >
              Submit 2
            </span>
          </button>
        </div>
        <div
          id="disabled-search-search-2"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <button
            className="usa-button-disabled"
            disabled="true"
            id="disabled-search-button-search-2"
            title="search button disabled"
            type="submit"
          >
            <span
              className="usa-search-submit-text usa-button-disabled"
            >
              Submit 2
            </span>
          </button>
          <span
            className="alert-text"
          >
            Search is disabled
          </span>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`SearchBarComponent small matches snapshot 1`] = `
<div
  className="usa-search usa-search-small searchbar"
>
  <div
    className="usa-grid-full"
    role="search"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="usa-grid-full label-input-wrapper"
      >
        <label
          className="usa-sr-only"
          htmlFor="search-2"
        >
          Label
        </label>
        <input
          disabled={false}
          id="search-2"
          name="search"
          onChange={[Function]}
          placeholder={null}
          type="search"
          value=""
        />
        <div
          id="enabled-search-search-2"
        >
          <button
            className=""
            disabled={false}
            id="enabled-search-button-search-2"
            title="submit search"
            type="submit"
          >
            <span
              className="usa-search-submit-text"
            />
            <span
              className="usa-sr-only"
            >
              Search
            </span>
          </button>
        </div>
        <div
          id="disabled-search-search-2"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <button
            className="usa-button-disabled"
            disabled="true"
            id="disabled-search-button-search-2"
            title="search button disabled"
            type="submit"
          >
            <span
              className="usa-search-submit-text usa-button-disabled"
            />
          </button>
          <span
            className="alert-text"
          >
            Search is disabled
          </span>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`SearchBarComponent small matches snapshot when showClear === true 1`] = `
<div
  className="usa-search usa-search-small searchbar"
>
  <div
    className="usa-grid-full"
    role="search"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="usa-grid-full label-input-wrapper"
      >
        <label
          className="usa-sr-only"
          htmlFor="search-2"
        >
          Label
        </label>
        <span
          className="text-input-wrapper"
        >
          <input
            disabled={false}
            id="search-2"
            name="search"
            onChange={[Function]}
            placeholder={null}
            type="search"
            value=""
          />
          <FontAwesome
            name="search"
          />
        </span>
        <div
          id="enabled-search-search-2"
        >
          <button
            className=""
            disabled={false}
            id="enabled-search-button-search-2"
            title="submit search"
            type="submit"
          >
            <span
              className="usa-search-submit-text"
            />
            <span
              className="usa-sr-only"
            >
              Search
            </span>
          </button>
        </div>
        <div
          id="disabled-search-search-2"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <button
            className="usa-button-disabled"
            disabled="true"
            id="disabled-search-button-search-2"
            title="search button disabled"
            type="submit"
          >
            <span
              className="usa-search-submit-text usa-button-disabled"
            />
          </button>
          <span
            className="alert-text"
          >
            Search is disabled
          </span>
        </div>
      </div>
    </form>
  </div>
</div>
`;
